### PR TITLE
Add Proxy Upstreams to Service Definition

### PR DESCRIPTION
**Pull request by banks**
**Note this is a resubmit of #4630** which got broken by some bad git-fu while rebasing onto new f-envoy branch and now I can't re-open.

----

This PR adds Proxy Upstream definitions to be first-class within service registrations in two separate places.

The motivation is that as we integrate Envoy and other proxies, they all need ways to express and centrally manage this configuration (i.e. in catalog not just local agent state) so it needs to become a first class part of the service definitions.

The upstreams for managed proxies can now be specified directly in the `connect.proxy.upstreams` stanza rather than being further nested inside the opaque map `connect.proxy.config` as part of the _application_ service registration.

In addition, "unmanaged" proxies can now register themselves with `Kind = "connect-proxy"` and with an explicit set of Upstreams as well as some other common proxy configuration.

The full details of this are documented in the doc changes inline here.

## Deprecations

This PR deprecates two things:
 1. Specifying upstreams for managed proxies in the service definitions opaque _config_ map
 2. Specifying an unmanaged proxy target in the `ProxyDestination` field

In both cases we gracefully accept the old value for now and translate it to the new setup. We should document these as deprecated though and remove them later after 1.3 release.

## Testing

It turns out this apparently simple change was a huge pain given our current baggage around chancing config structs and supporting multiple formats and legacy key translations from snake to camel case etc.

I methodically considered and added tests for all of the following cases:

- Agent Services endpoint response
- Agent Service endpoint response
- Agent Register endpoint
   - Unmanaged deprecated field
   - Unmanaged new fields
   - Managed deprecated upstreams
   - Managed new
- Agent Proxy config endpoint response
- Catalog Register
   - Unmanaged deprecated field
   - Unmanaged new fields
   - Managed deprecated upstreams
   - Managed new
- Catalog Node endpoint response
- Catalog Service endpoint response
- API package methods for ALL of the above endpoints/cases
- Config file parsing for service/services in JSON/HCL
- Config file translations from deprecated to new fields

Even with all those tests, I still wasn't confident so I wrote an external sanity check script: https://github.com/banks/consul-upstream-config-e2e

This found a whole bunch of additional issues! For now I've left it out of our actual test suite because for each issue it did find I was able to add additional cases to the actual tests we have that covered them before fixing so I don't think it adds any benefit to include this script in the repo as it is. We _will_ look at incorporating more cases like this though into our existing functional/e2e test suite though as I think this showed there is value for significant config changes (especially in service definitions which touch so many parts of the API and config code) to validate all the options.

Even with that done, several minor issues were found and fixed while writing up documentation for the endpoints with additional tests added:
 - Catalog service responses must still include deprecated `ServiceProxyDestination` which is back-populated from the new fields. This is probably not used by anyone but it felt odd to do a half-baked job of backward compatibility in the API and would technically be a breaking change to remove it.
 - A few optional empty fields should be omitted in some responses but were not

## Documentation

Should all be in separate commit 6b5812db3 for easy review.

## TODO

 - [ ] Fix documentation to refer to 1.3 as the version that this will be released in - it states 1.2.3 in several places currently.

### CODE REVIEW

**Comment by banks** [comment]
^ force push/amend just to trigger CI which I think I've enabled for this branch now.

**Comment by mkeeler** [comment]
Is the plan to remove the ServiceProxyDestination output from the API when connect exits beta?

**Comment by banks** [comment]
I guess we probably would leave it for at least another major release cycle given that this will likely only be out at best a few weeks before GA.

File: website/source/docs/connect/proxies.html.md [hunk 1]
```diff
@@ -134,8 +132,17 @@ static port will be able to masquerade as the source service ("web" in the
 example above). You must either trust any loopback access on that port or
 use namespacing techniques provided by your operating system.
 
-For full details of the configurable options available see the [built-in proxy
-configuration reference](/docs/connect/configuration.html#built-in-proxy-options).
+-> **Deprecation Note:** versions 1.2.0 to 1.2.3 required specifying `upstreams`
```
Comment (mkeeler) [inline website/source/docs/connect/proxies.html.md]: Just a note but version numbers might need updating here.

File: website/source/docs/connect/proxies.html.md [hunk 2]
```diff
@@ -187,6 +193,43 @@ with Connect. After natively integrating, the HTTP API or
...
+  prepared query to route connect to.
+* `local_bind_port` `int: <required>` - Specifies the port to bind a local
+  listener to for the application to make outbound connections to this upstream.
+* `local_bind_address` `string: <optional>` - Specifies the address to bind a
+  local listener to for the application to make outbound connections to this
+  upstream. Defaults to `127.0.0.1`.
+* `destination_type` `string: <optional>` - Speficied the type of discovery
+  query to use to find an instance to connect to. Valid values are `service` or
+  `prepared_query`. Defaults to `service`.
+* `datacenter` `string: <optional>` - Specifies the datacenter to issue the
```
Comment (mkeeler) [inline website/source/docs/connect/proxies.html.md]: Should this be destination_datacenter or is the JSON above wrong?

File: agent/agent_endpoint.go [hunk 3]
```diff
@@ -554,6 +595,32 @@ func (s *HTTPServer) AgentRegisterService(resp http.ResponseWriter, req *http.Re
...
+						proxyMap = pMap
+					}
+					if pMap, ok := connectMap["proxy"].(map[string]interface{}); ok {
+						proxyMap = pMap
+					}
+					if proxyMap != nil {
+						config.TranslateKeys(proxyMap, map[string]string{
+							"exec_mode": "ExecMode",
+						})
+						translateUpstreams(proxyMap)
```
[the remark this answers is not in the document]
Comment (banks) [reply agent/agent_endpoint.go]: Nope this this the "new" place for managed proxies!

[the remark this answers is not in the document]
Comment (banks) [reply agent/agent_endpoint.go]: _Now_ it's deprecated 😉 

Going to leave it here for this PR for reasons mentioned before. I'll also add Deprecation notices to all managed proxy config and components in a later PR.

**Pull request merged**